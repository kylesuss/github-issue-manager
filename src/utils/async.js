import superagent from 'superagent'

export const get = (url, options) => {
  return new Promise((resolve, reject) => {
    const request = superagent.get(url)

    Object.keys(options).forEach((key) => request.set(key, options[key]))

    request.end((error, response) => {
      if (error) { return reject(error) }
      resolve(response)
    })
  })
}
