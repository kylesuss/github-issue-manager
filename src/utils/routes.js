export const buildRoute = (path, variables) => {
  const pathParts = path.split('/')
  const dynamicParts = pathParts.filter(part => part.match(/^:/))

  return dynamicParts.reduce((path, dynamicPart) => (
    path.replace(dynamicPart, variables[dynamicPart.slice(1)] || dynamicPart)
  ), path)
}
