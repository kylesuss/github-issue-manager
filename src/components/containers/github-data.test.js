import React from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import flushPromises from 'test/flush-promises'
import * as async from 'utils/async'
import withGithubData from './github-data'

jest.mock('utils/async', () => ({
  get: jest.fn(() => new Promise((resolve) => resolve({ body: ':responseBody' })))
}))

const TEST_URL = ':testUrl'
const WrappedComponent = ({ data }) => (
  <div>
    <span>Error: {data.error}</span>
    <span>Response: {data.response}</span>
    <span>Loading: {data.isLoading.toString()}</span>
  </div>
)

WrappedComponent.propTypes = {
  data: PropTypes.shape({
    isLoading: PropTypes.bool.isRequired,
    response: PropTypes.string
  }),
  error: PropTypes.string,
  response: PropTypes.string,
}

test('it renders the wrapped component before the request is done', async () => {
  const ComponentWithGithubData = withGithubData(TEST_URL)(WrappedComponent)
  const wrapper = mount(<ComponentWithGithubData />)
  expect(toJson(wrapper)).toMatchSnapshot()
})

test('it renders the wrapped component with a response', async () => {
  const ComponentWithGithubData = withGithubData(TEST_URL)(WrappedComponent)
  const wrapper = mount(<ComponentWithGithubData />)

  await flushPromises()
  wrapper.update()

  expect(toJson(wrapper)).toMatchSnapshot()
})

test('it renders the wrapped component with an error', async () => {
  async.get.mockReturnValueOnce(new Promise((_, reject) => reject(':error')))
  const ComponentWithGithubData = withGithubData(TEST_URL)(WrappedComponent)
  const wrapper = mount(<ComponentWithGithubData />)

  await flushPromises()
  wrapper.update()

  expect(toJson(wrapper)).toMatchSnapshot()
})
