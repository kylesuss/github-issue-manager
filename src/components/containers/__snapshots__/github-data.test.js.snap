// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders the wrapped component before the request is done 1`] = `
<GitHubData>
  <WrappedComponent
    data={
      Object {
        "error": null,
        "isLoading": true,
        "response": null,
      }
    }
  >
    <div>
      <span>
        Error: 
      </span>
      <span>
        Response: 
      </span>
      <span>
        Loading: 
        true
      </span>
    </div>
  </WrappedComponent>
</GitHubData>
`;

exports[`it renders the wrapped component with a response 1`] = `
<GitHubData>
  <WrappedComponent
    data={
      Object {
        "error": null,
        "isLoading": false,
        "response": ":responseBody",
      }
    }
  >
    <div>
      <span>
        Error: 
      </span>
      <span>
        Response: 
        :responseBody
      </span>
      <span>
        Loading: 
        false
      </span>
    </div>
  </WrappedComponent>
</GitHubData>
`;

exports[`it renders the wrapped component with an error 1`] = `
<GitHubData>
  <WrappedComponent
    data={
      Object {
        "error": ":error",
        "isLoading": false,
        "response": null,
      }
    }
  >
    <div>
      <span>
        Error: 
        :error
      </span>
      <span>
        Response: 
      </span>
      <span>
        Loading: 
        false
      </span>
    </div>
  </WrappedComponent>
</GitHubData>
`;
